src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -122,50 +109,33 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
             Func<Solution, ValueTask<T>> implementation,
             CancellationToken cancellationToken)
         {
-            return RunWithSolutionAsync(assetProvider, solutionChecksum, workspaceVersion: -1, fromPrimaryBranch: false, implementation, cancellationToken);
+            return RunWithSolutionAsync(assetProvider, solutionChecksum, workspaceVersion: -1, updatePrimaryBranch: false, implementation, cancellationToken);
         }
 
         private async ValueTask<(Solution solution, T result)> RunWithSolutionAsync<T>(
             AssetProvider assetProvider,
             Checksum solutionChecksum,
             int workspaceVersion,
-            bool fromPrimaryBranch,
+            bool updatePrimaryBranch,
             Func<Solution, ValueTask<T>> implementation,
             CancellationToken cancellationToken)
         {
-            // Trivial case.  See if the checksum being asked for actually corresponds to this workspace's current
-            // solution.  If so, just use that directly:
-
-            var currentSolution = this.CurrentSolution;
-            var currentSolutionChecksum = await currentSolution.State.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
-            if (currentSolutionChecksum == solutionChecksum)
-                return (currentSolution, await implementation(currentSolution).ConfigureAwait(false));
-
-            // Next, look in our caches to see if we can find the item.
-
-            // !!!CRITICAL!!! Ensure the moment we get the refCountedLazySolution we put it in a using-block.  That way
-            // if anything causes us to abort/cancel, we'll reduce the incremented ref-count that GetLazySolutionAsync
-            // caused.  Do NOT add anything between this line and the next.
```

Comment:
i moved away from async disposal.  all gates are short lived and it dramatically simplifies things this way.